server/core/basic_cluster.go

```diff
@@ -317,6 +374,12 @@ func (bc *BasicCluster) PreCheckPutRegion(region *RegionInfo) (*RegionInfo, erro
 func (bc *BasicCluster) PutRegion(region *RegionInfo) []*RegionInfo {
 	bc.Lock()
 	defer bc.Unlock()
+	if bc.SelectConf.NewRegionFirst {
```

Comment:
Could you make this function as a `Hook`, and extract this logic looks like  an plugin.